rpcserver.go

```diff
@@ -5373,3 +5445,75 @@ func (r *rpcServer) BakeMacaroon(ctx context.Context,
 
 	return resp, nil
 }
+
+// FundingStateStep is an advanced funding related call that allows the caller
+// to either execute some preparatory steps for a funding workflow, or manually
+// progress a funding workflow. The primary way a funding flow is identified is
+// via its pending channel ID. As an example, this method can be used to
+// specify that we're expecting a funding flow for a particular pending channel
+// ID, for which we need to use specific parameters. Alternatively, this can be
+// used to interactively drive PSBT signing for funding for partially complete
+// funding transactions.
+func (r *rpcServer) FundingStateStep(ctx context.Context,
+	in *lnrpc.FundingTransitionMsg) (*lnrpc.FundingStateStepResp, error) {
+
+	switch {
+
+	// If this is a message to register a new shim that is an external
+	// channel point, then we'll contact the wallet to register this new
+	// shim. A user will use this method to register a new channel funding
+	// workflow which has already been partially negotiated outside of the
+	// core protocol.
+	case in.GetShimRegister() != nil &&
+		in.GetShimRegister().GetChanPointShim() != nil:
+
+		rpcShimIntent := in.GetShimRegister().GetChanPointShim()
+
+		// Using the rpc shim as a template, we'll construct a new
+		// chanfunding.Assembler that is able to express proper
+		// formulation of this expected channel.
+		shimAssembler, err := newFundingShimAssembler(
+			rpcShimIntent, false,
+		)
+		if err != nil {
+			return nil, err
+		}
+		shimIntent, err := shimAssembler.ProvisionChannel(nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Once we have the intent, we'll register it with the wallet.
+		// Once we receive an incoming funding request that uses this
+		// pending channel ID, then this shim will be dispatched in
+		// place of our regular funding workflow.
+		var pendingChanID [32]byte
+		copy(pendingChanID[:], rpcShimIntent.PendingChanId)
+		err = r.server.cc.wallet.RegisterFundingIntent(
```

Comment:
as noted in the other review, since the canned assembler can also handle responder requests, I think we could register the `shimAssembler` with the wallet here instead of the intent, such that the logic in the reservation code would have one less code path to take.

Reply:
Which code path? This is only intended to mark an incoming request for interception. We'll need to track the set of requests to be intercepted _somewhere_, with the other alternative being in the funding manager itself. I prefer doing it in the wallet as it allows all the logic to be tested in a single package, with the funding manager being mostly unaware of it all (notice how it's hardly changed in this diff and the prior PR, that was intended).

Reply:
Suggestion was to just register the `shimAssembler` instead, but now that the other PR is merged this is no longer an option.